370-auth-ui/oidc.md

```diff
@@ -0,0 +1,455 @@
+
+The OpenID Connect discovery document can be retrieved from:
+
+```
+https://{auth_gear_endpoint}/.well-known/openid-configuration
+```
+
+The OAuth Authorization Server Metadata can be retrieved from:
+
+```
+https://{auth_gear_endpoint}/.well-known/oauth-authorization-server
+```
+
+The following sessions list out the metadata values that Skygear Auth supports.
+
+#### issuer
+
+No difference from the spec.
+
+#### authorization_endpoint
+
+No difference from the spec.
+
+#### token_endpoint
+
+No difference from the spec.
+
+#### userinfo_endpoint
+
+No difference from the spec.
+
+#### jwks_uri
+
+No difference from the spec.
+
+### scopes_supported
+
+The values will be `["openid", "offline_access"]`.
+
+#### response_types_supported`
+
+The values will be `["code", "none"]`.
+
+#### grant_types_supported
+
+The values will be `["authorization_code", "refresh_token"]`.
+
+#### subject_types_supported
+
+The values will be `["public"]`.
+
+#### id_token_signing_alg_values_supported
+
+`RS256` is supported.
+
+#### claims_supported
+
+The values will be `["sub", "iss", "aud", "exp", "iat"]`.
+
+### code_challenge_methods_supported
+
+The values will be `["plain", "S256"]`
+
+### revocation_endpoint
+
+No difference from the spec. See [Revocation Request](https://tools.ietf.org/html/rfc7009#section-2.1) for detail.
+
+## Signing ID token
+
+ID token signing key will be generated per app during app creation and store in App Config. For config compatibility, keys is an array, but Auth Gear only supports one key in this stage. Keys length will be checked by JSON schema.
+
+```yaml
+app_config:
+  oidc:
+    keys:
+    - kid: key_1
+      public_key: ""
+      private_key: ""
+```
+
+## Configurations
+
+### Sample configuration
+
+```yaml
+app_config:
+  clients:
+  - client_id: XXX
+    client_secret: XXX
+    refresh_token_lifetime: 86400
+    access_token_lifetime: 1800
+    session_idle_timeout_enabled: true
+    session_idle_timeout: 300
+    client_name: Mobile App
+    redirect_uris:
+    - "https://example.com"
+    logo_uri: "https://example.com/logo.png"
+    grant_types:
+    - "authorization_code"
+    - "refresh_token"
+    response_types:
+    - "code"
+```
+
+### Parameters
+
+#### OIDC client Metadata
+
+Some parameters are defined in OIDC, see [ClientMetadata](https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata).
+
+They are
+
+- `client_name`
+- `redirect_uris`
+- `login_uri`
+- `grant_types`
+- `response_types`
+
+#### Skygear-specific
+
+- `client_id`: OIDC client id.
+- `client_secret`: OIDC client secret, used in RP integration.
+- `access_token_lifetime`: OIDC access token lifetime in seconds, default to 1800.
+- `refresh_token_lifetime`: OIDC refresh token lifetime in seconds, default to max(access_token_lifetime, 86400). Must greater than or equal to `access_token_lifetime`.
+- `session_idle_timeout_enabled`: Indicate whether OIDC session idle timeout is enabled, default to `false`.
```

Comment:
Session idle timeout is not client-specific, so the config can be removed here.

Reply:
Yes, you are right. I removed the session idle config.

But would like to clarify the session idle behavior. So if user access any api with Idp session / refresh token / access token, the last access at of session will be updated. If last access at pass the session idle timeout, all those session and tokens will be invalidate?

Reply:
Allow me to clarify:
- IdP session: have session cookie
- Offline grant: have refresh token & access tokens

IdP sessions and offline grants are independent of each other (even if user logged out from IdP session, the offline grant is still valid). Session idle behavior only applies to IdP sessions.

Reply:
Should we have idle timeout for refresh token? Access Token should be kept simple and does not have idle timeout.

Reply:
We do not plan to expose manual refresh function in SDK, and it'll use refresh token regularly according to the expiry of access token, so I think it's mostly meaningless.

Reply:
How should a third party banking mobile application configure the settings if it wants idle timeout?

Reply:
I suppose the use case is that refresh token should expire soon if the app is inactive (i.e. in background). In this case, we can separate refresh token lifetime into two config: one for maximum lifetime of the offline grant, one for the lifetime of individual refresh token.

ref: http://docs.identityserver.io/en/stable/topics/refresh_tokens.html

Reply:
So the refresh token itself is also refreshed?

Reply:
Yes, it is a possible solution. However, I think we don't need to include these mechanism for now.

Reply:
So the conclusion is that both access token and refresh token has maximum lifetime only. Refresh token is static and never refresh.

Reply:
thanks @louischan-oursky @kiootic :)